pmlc/conversion/gpu_to_comp/gpu_to_comp.cc

```diff
@@ -105,121 +113,173 @@ RewriteLaunchFunc::matchAndRewrite(gpu::LaunchFuncOp op,
         "Expected containing function to supply an execution device");
     return mlir::failure();
   }
-  mlir::Value device = func.getArgument(0);
   // Create execution environment.
+  mlir::Value device = func.getArgument(0);
   auto execEnvOp =
       rewriter.create<comp::CreateExecEnv>(loc, execEnvType, device);
   mlir::Value execEnv = execEnvOp.getResult();
+  // Collect consecutive gpu.launch_func Ops starting from current op.
+  std::vector<gpu::LaunchFuncOp> launchOps;
+  if (mlir::failed(getConsecutiveOps<gpu::LaunchFuncOp>(op, launchOps)))
+    return mlir::failure();
+
+  std::map<mlir::Value, mlir::Value, valueComparator> bufferPool;
+  std::vector<std::vector<mlir::Value>> newOperands(launchOps.size());
   // Allocate memory on device for memory operands.
-  std::vector<mlir::Value> newOperands;
-  if (mlir::failed(allocateDeviceMemory(rewriter, loc, execEnv, op.operands(),
-                                        newOperands)))
+  if (mlir::failed(allocateDeviceMemory(rewriter, loc, execEnv, launchOps,
+                                        newOperands, bufferPool)))
     return mlir::failure();
-  mlir::Value funcEvent;
-  if (mlir::failed(createScheduleFuncOp(rewriter, loc, execEnv, op, newOperands,
-                                        funcEvent)))
+  std::vector<mlir::Value> funcEvents;
+  // Create schedule functions within the same execution environment.
+  if (mlir::failed(createScheduleFuncOps(rewriter, loc, execEnv, launchOps,
+                                         newOperands, funcEvents)))
     return mlir::failure();
   // Read device memory back to host.
-  if (mlir::failed(readDeviceMemory(rewriter, loc, execEnv, op.operands(),
-                                    newOperands, funcEvent)))
+  if (mlir::failed(readDeviceMemory(rewriter, loc, execEnv, funcEvents,
+                                    newOperands, bufferPool)))
     return mlir::failure();
   // Deallocate device memory.
-  if (mlir::failed(deallocateDeviceMemory(rewriter, loc, execEnv, op.operands(),
-                                          newOperands)))
+  if (mlir::failed(deallocateDeviceMemory(rewriter, loc, execEnv, bufferPool)))
     return mlir::failure();
   // Destroy execution environment.
   rewriter.create<comp::DestroyExecEnv>(loc, execEnv);
-  // Remove original launch operation.
-  rewriter.eraseOp(op.getOperation());
+  // Remove original launch operations.
+  for (auto launchOp : launchOps) {
+    rewriter.eraseOp(launchOp.getOperation());
+  }
+  return mlir::success();
+}
 
+template <typename T>
+mlir::LogicalResult
+RewriteLaunchFunc::getConsecutiveOps(T op, std::vector<T> &ops) const {
+  // Collect consecutive ops of type T starting from a given op.
+  auto block = op.getOperation()->getBlock();
+  for (auto itOp = block->begin(); itOp != block->end(); itOp++) {
+    if (mlir::isa<T>(itOp) && op == mlir::cast<T>(*itOp)) {
```

Comment:
Again, it might be more clear to do this in two loops:  1) find `op` in the block and 2) find all consecutive `ops` with same type T following `op`.  There may be an MLIR standard way to do step 1 as well.